### PR TITLE
Fix pyerial issue and add support for Python 3

### PR DESCRIPTION
**Pull request by Mani-Sadhasivam**
This PR fixes the Pyserial issue observed on most of the platforms and also adds Python 3 support without breaking the Python 2.7 compatibility.

Issue: https://github.com/96boards/documentation/issues/506

### CODE REVIEW

**Comment by hzhuang1** [comment]
Why do you need to extract l-loader.bin from AOSP images?

**Comment by Mani-Sadhasivam** [comment]
@hzhuang1 According to the [recovery instructions](https://github.com/96boards/documentation/blob/master/consumer/hikey970/installation/board-recovery.md#run-the-script-to-initially-prepare-fastboot), we need `l-loader.bin` for executing this script. Please see the doumentation issue I have referenced in the PR description.

Thanks,
Mani

**Comment by hzhuang1** [comment]
l-loader.bin should be generated. Why do you want to copy l-loader.bin? I think user should generate it by himself.

**Comment by Mani-Sadhasivam** [comment]
@hzhuang1 There is no information in the [recovery guide](https://github.com/96boards/documentation/blob/master/consumer/hikey970/installation/board-recovery.md) on how to generate the `l-loader.bin` file! Can you please clarify?

File: hisi-idt.py [hunk 1]
```diff
@@ -128,8 +129,8 @@ def sendstartframe(self):
         self.s.timeout =0.01
         data = array.array('B', self.startframe[self.chip]).tostring()
         crc = self.calc_crc(data)
-        data += chr((crc >> 8)&0xff)
-        data += chr(crc&0xff)
+        data += struct.pack("B",((crc >> 8) & 0xff))
+        data += struct.pack("B",((crc) & 0xff))
```
Comment (daniel-thompson) [inline hisi-idt.py]: This is bizzare. The original array.array(...).tostring() is an odd way to construct a bytearray and using struct.pack() is insert things into a bytearray is also pretty odd.

Should be:

data = bytestring(self.startframe[self.chip])
crc = self.calc_crc(data)
data.append((crc >> 8) & 0xff)
data.append(crc & 0xff)

I also note that the above is a repeated pattern in the code... it should be pulled out into an add_crc() function.

File: hisi-idt.py [hunk 2]
```diff
@@ -100,7 +101,7 @@ def __del__(self):
 
     def calc_crc(self, data, crc=0):
         for char in data:
-            crc = ((crc << 8) | ord(char)) ^ self.crctable[(crc >> 8) & 0xff]
+            crc = ((crc << 8) | ord(chr(char))) ^ self.crctable[(crc >> 8) & 0xff]
```
Comment (daniel-thompson) [inline hisi-idt.py]: chr() converts an integer to a character; ord() converts a character to an integer. If we know char is < 256 (which we do) then ord(chr(char)) is a no-op.

File: hisi-idt.py [hunk 3]
```diff
@@ -142,37 +143,34 @@ def sendheadframe(self,length,address):
...
         self.headframe[self.chip][11] = (address)&0xff
-
         data = array.array('B', self.headframe[self.chip]).tostring()
         crc = self.calc_crc(data)
-
-        data += chr((crc >> 8)&0xff)
-        data += chr(crc&0xff)
-
+        data += struct.pack("B",((crc >> 8) & 0xff))
+        data += struct.pack("B",((crc) & 0xff))
```
Comment (daniel-thompson) [inline hisi-idt.py]: As above.

File: hisi-idt.py [hunk 4]
```diff
@@ -117,7 +118,7 @@ def sendframe(self, data, loop):
             try:
                 ack = self.s.read()
```
Comment (daniel-thompson) [inline hisi-idt.py:118]: Perhaps: ack = bytearray(self.s.read())

File: hisi-idt.py [hunk 5]
```diff
@@ -117,7 +118,7 @@ def sendframe(self, data, loop):
...
                 if len(ack) == 1:
-                    if ack == chr(0xaa):
+                    if ack == struct.pack("B", 0xaa):
```
Comment (daniel-thompson) [inline hisi-idt.py]: Perhaps: if ack[0] == 0xaa:

File: hisi-idt.py [hunk 6]
```diff
@@ -142,37 +143,34 @@ def sendheadframe(self,length,address):
...
 
 
     def senddataframe(self,seq,data):
         self.s.timeout = 0.15
-        head = chr(0xDA)
-        head += chr(seq&0xFF)
-        head += chr((~seq)&0xFF)
+        head = struct.pack("B",0xda)
+        head += struct.pack("B",(seq & 0xff))
+        head += struct.pack("B",((~seq) & 0xff))
```
Comment (daniel-thompson) [inline hisi-idt.py]: Again... why use struct.pack just to create a bytearray

Either:

head = bytearray((0xda, seq & 0xff, (~seq & 0xff))

Or:

head = b'0xda'
head.append(seq & 0xff)
head.append((~seq) & 0xff)

File: hisi-idt.py [hunk 7]
```diff
@@ -142,37 +143,34 @@ def sendheadframe(self,length,address):
...
 
         data = head + data
 
         crc = self.calc_crc(data)
-        data += chr((crc >> 8)&0xff)
-        data += chr(crc&0xff)
+        data += struct.pack("B",((crc >> 8) & 0xff))
+        data += struct.pack("B",((crc) & 0xff))
```
Comment (daniel-thompson) [inline hisi-idt.py]: As above.

File: hisi-idt.py [hunk 8]
```diff
@@ -142,37 +143,34 @@ def sendheadframe(self,length,address):
...
 
         self.sendframe(data,32)
 
     def sendtailframe(self,seq):
-        data = chr(0xED)
-        data += chr(seq&0xFF)
-        data += chr((~seq)&0xFF)
+        data = struct.pack("B",0xed)
+        data += struct.pack("B",(seq & 0xff))
+        data += struct.pack("B",((~seq) & 0xff))
```
Comment (daniel-thompson) [inline hisi-idt.py]: As above.

Comment (daniel-thompson) [inline hisi-idt.py]: As above

**Review by daniel-thompson — CHANGES_REQUESTED** [review]
IMHO the elimination of python2 support is gratuitous.

When the code is modified to use simpler methods to manipulate bytearrays is should become obvious that simply wrapping the two calls to read (in sendframe() and download() in bytearray(thing.read()) will make the code compatible with python2.

**Comment by hzhuang1** [comment]
@Mani-Sadhasivam You can get the reference of hikey960.mk. I prefer you can copy it to hikey970.mk.

By the way, do we need to upgrade it to python3? There're a few scripts written in python2.7. It seems not good to maintain the scripts of both python2 and python3.

File: hisi-idt.py [hunk 9]
```diff
@@ -85,7 +86,7 @@ class bootdownload(object):
 
     def __init__(self,chiptype,serialport):
         try:
-            self.s = serial.Serial(port=serialport, baudrate=115200, timeout=1)
+            self.s = serial.Serial(dsrdtr=True, rtscts=True, port=serialport, baudrate=115200, timeout=1)
```
Comment (daniel-thompson) [inline hisi-idt.py:88]: Hmnnn... looks like I missed this one yesterday (I typed it in... maybe I deleted it by mistake). Either way...

This change doesn't seem to have anything to do with python3; it is a bug fix to make pyserial work on hikey960/970 and should be a separate patch.

File: hisi-idt.py (hunk 1 above)
Comment (Mani-Sadhasivam) [reply hisi-idt.py]: Sure. Will fix it in next revision. I'm sure that you mean, `bytearray` not `bytestring`.

File: hisi-idt.py (hunk 2 above)
Comment (Mani-Sadhasivam) [reply hisi-idt.py]: Agree.

File: hisi-idt.py (hunk 4 above)
Comment (Mani-Sadhasivam) [reply hisi-idt.py:118]: I think this can be left as is. Please see the comment below

File: hisi-idt.py (hunk 5 above)
Comment (Mani-Sadhasivam) [reply hisi-idt.py]: Better way would be

'if ack == b'\xaa':'

Since we receive bytes!

File: hisi-idt.py (hunk 6 above)
Comment (Mani-Sadhasivam) [reply hisi-idt.py]: Ack. Btw, `head = b'0xda'` will just be bytes. We need to convert it to bytearray as below for appending:

`head = bytearray(b'0xda')`

File: hisi-idt.py (hunk 8 above)
Comment (Mani-Sadhasivam) [reply hisi-idt.py]: Ack

Comment (Mani-Sadhasivam) [reply hisi-idt.py]: Ack

File: hisi-idt.py (hunk 9 above)
Comment (Mani-Sadhasivam) [reply hisi-idt.py:88]: Sure, will split the patches.

Thanks for your detailed review :)

**Comment by Mani-Sadhasivam** [comment]
@hzhuang1 Modified the PR to incorporate review by @daniel-thompson and also removed the commit for adding `l-loader.bin`. Will submit another PR for adding the `hikey970.mk` as you suggested.

File: hisi-idt.py (hunk 4 above)
Comment (daniel-thompson) [reply hisi-idt.py:118]: Couldn't find the comment below... but agreed anyway.

File: hisi-idt.py [hunk 10]
```diff
@@ -117,7 +117,7 @@ def sendframe(self, data, loop):
             try:
                 ack = self.s.read()
                 if len(ack) == 1:
-                    if ack == chr(0xaa):
+                    if ack == b'\xaa':
```
Comment (daniel-thompson) [inline hisi-idt.py]: Doesn't doing a string compare here make the len(ack) == 1 check pointless?

**Review by daniel-thompson — COMMENTED** [review]
Tiny nitpick but otherwise LGTM

File: hisi-idt.py (hunk 10 above)
Comment (Mani-Sadhasivam) [reply hisi-idt.py]: Agree. Will remove this

**Comment by Mani-Sadhasivam** [comment]
@daniel-thompson Removed the `if` condition checking the ack length.

**Review by daniel-thompson — APPROVED** [review]
LGTM

**Comment by jainrahul1** [comment]
@Mani-Sadhasivam  @daniel-thompson I am facing issue in recovery of my Hikey970 board,

Setup:- I flashed hikey with https://github.com/mengzhuo/hikey970-ubuntu-image/releases/tag/v1.0
Now i want to revert back to android so I am doing the recovery.
Switch1,2,3 are kept in ON state, I have download the android release and copied l-loader.bin to tool directory. 
cmd: sudo python hisi-idt.py -d /dev/ttyXRUSB0 --img1 ./sec_usb_xloader.img --img2 ./sec_usb_xloader2.img --img3 ./l-loader.bin

+----------------------+
(' Serial: ', '/dev/ttyXRUSB0')
(' Image1: ', './sec_usb_xloader.img')
(' Image2: ', './sec_usb_xloader2.img')
+----------------------+

('Sending', './sec_usb_xloader.img', '...')

It get stuck in Sending command. Could please advice what to do?

**Comment by Mani-Sadhasivam** [comment]
Hi @jainrahul1,

Your tty device seems suspicious. Usually, it will be `ttyUSBx` on Unix based systems. Can you please verify?

**Comment by Mani-Sadhasivam** [comment]
@jainrahul1 

`ttyXRUSB` refers to the Debug UART port, so you are connecting to the wrong Type C port. For recovering HiKey970, you need to use the other Type C port (between HDMI and USB ports).
